### PR TITLE
Added Gradient Descent [Python]

### LINKED ISSUE #345: Gradient Descent [Python]
#### This issue is about [Algo/DS Name](link to resource for the Algo/DS).

- [x] Issue Title is in form `Algo/DS Name [Language]`.
- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [x] This Algo/DS is worth adding to this repository.
- [x] Assign this issue to me (I want to work on this).
	- [x] I ensure that I am not already assigned to 2 or more issues.
- [x] I will send my PR only after I'm assigned to this issue by any maintainer.
- [x] I understand that I will be unassigned from this issue if I fail to send a PR for this within a week.

### PR DESCRIPTION
**Pull request by Prakash2403**
Fixes #345 

#### By submitting this pull request I confirm I've read and complied with the below declarations.

- [x] I have read the [Contribution guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md) and I am confident that my PR reflects them.
- [x] I have followed the [coding guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#cs) for this project.
- [x] My code follows the [skeleton code structure](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#sample).
- [x] This pull request has a descriptive title. For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.
- [x] This pull request will be closed if I fail to update it even once in a continuous time span of 7 days.

### CODE REVIEW

File: gradient_descent/gradient_descent.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    return calculate_hypothesis_value(example_no, data_set) - output(example_no, data_set)
+
+
+def _hypothesis_value(data_input_tuple):
+    """
+    Calculates hypothesis function value for a given input
+    :param data_input_tuple: Input tuple of a particular example
+    :return: Value of hypothesis function at that point.
+    Note that parameter input value is fixed as 1.
+    Also known as 'biased input' inn ML terminology and the parameter associated with it
```
Comment (yashLadha) [inline gradient_descent/gradient_descent.py]: change `inn` to `in`

**Comment by Prakash2403** [comment]
@yashLadha Please review again.

File: gradient_descent/gradient_descent.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+def run_gradient_descent():
+    global parameter_vector
+    # Tune these values to set a tolerance value for predicted output
+    absolute_error_limit = 0.000002
+    relative_error_limit = 0
+    j = 0
+    while True:
+        j = j+1
+        temp_parameter_vector = [0, 0, 0, 0]
+        for i in range(0, len(parameter_vector)):
```
Comment (aashutoshrathi) [inline gradient_descent/gradient_descent.py]: `for i in range(0, len(parameter_vector)):` can be changed to `for i in range(len(parameter_vector)):`

File: gradient_descent/gradient_descent.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        if numpy.allclose(parameter_vector, temp_parameter_vector,
+                          atol=absolute_error_limit, rtol=relative_error_limit):
+            break
+        parameter_vector = temp_parameter_vector
+    print("Number of iterations:", j)
+
+
+def test_gradient_descent():
+    for i in range(len(test_data)):
+        print("Actual output value:",  output(i, 'test'))
```
Comment (aashutoshrathi) [inline gradient_descent/gradient_descent.py]: Change `print("Actual output value:",  output(i, 'test'))` to `print("Actual output value:", output(i, 'test'))`

File: gradient_descent/gradient_descent.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+
+def run_gradient_descent():
+    global parameter_vector
+    # Tune these values to set a tolerance value for predicted output
+    absolute_error_limit = 0.000002
+    relative_error_limit = 0
+    j = 0
+    while True:
+        j = j+1
```
Comment (aashutoshrathi) [inline gradient_descent/gradient_descent.py]: can be changed to `j += 1`

File: gradient_descent/gradient_descent.py (hunk 3 above)
Comment (Prakash2403) [inline gradient_descent/gradient_descent.py]: It is mentioned in PEP8 guidelines to have a space between `,` and next word.

**Comment by Prakash2403** [comment]
@aashutoshrathi Please review again.

File: gradient_descent/gradient_descent.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    :param data_input_tuple: Input tuple of a particular example
+    :return: Value of hypothesis function at that point.
+    Note that parameter input value is fixed as 1.
+    Also known as 'biased input' in ML terminology and the parameter associated with it
+    is known as 'biased parameter'.
+    """
+    hyp_val = 0
+    for i in range(len(parameter_vector) - 1):
+        hyp_val = hyp_val + data_input_tuple[i]*parameter_vector[i+1]
+    hyp_val = hyp_val + 1*parameter_vector[0]
```
Comment (aashutoshrathi) [inline gradient_descent/gradient_descent.py]: @Prakash2403 Sir , Why is this `1*parameter_vector[0]` ?

Comment (aashutoshrathi) [inline gradient_descent/gradient_descent.py]: Line 35 and 36 can be changed to `+=` format.

Comment (Prakash2403) [inline gradient_descent/gradient_descent.py]: As I have mentioned in comments, there is always a biased input in Artificial Neural Networks, whose value is fixed as 1. I wanted to explicitly mention this fact in the code too, that's why I have written `1*parameter_vector[0]` instead of `parameter[0]`.

But now, I guess it's better to keep it `parameter[0]`. Thanks for pointing it out.

**Comment by Prakash2403** [comment]
@aashutoshrathi Please review again.

File: gradient_descent/gradient_descent.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+
+def _hypothesis_value(data_input_tuple):
+    """
+    Calculates hypothesis function value for a given input
+    :param data_input_tuple: Input tuple of a particular example
+    :return: Value of hypothesis function at that point.
+    Note that there is an 'biased input' whose value is fixed as 1.
+    It is not explicitly mentioned in input data.. But, ML hypothesis functions use it.
+    So, we have to take care of it separately. Line 36 takes care of it.
```
Comment (aashutoshrathi) [inline gradient_descent/gradient_descent.py]: Remove Trailing Whitespaces

Comment (Prakash2403) [inline gradient_descent/gradient_descent.py]: Where is the issue exactly?

Comment (aashutoshrathi) [inline gradient_descent/gradient_descent.py]: At the end of Line 31

Comment (Prakash2403) [inline gradient_descent/gradient_descent.py]: Got it, thanks.

**Comment by Prakash2403** [comment]
@aashutoshrathi Please review again.

**Comment by yashLadha** [comment]
Just update your branch

File: gradient_descent/gradient_descent.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    hypothesis_output = _hypothesis_value(test_data,
+                                          parameter_vector=parameter_vector)
+    num_examples = len(test_output)
+    plt.stem(range(num_examples), actual_output, markerfmt='go', label='Actual Output')
+    plt.stem(range(num_examples), hypothesis_output, label='Hypothesis Output')
+    plt.xlabel('Test case')
+    plt.ylabel('Output Values')
+    plt.xlim([-1, 7])
+    plt.legend()
+    plt.show()
```
Comment (yashLadha) [inline gradient_descent/gradient_descent.py]: Won't the program end. If u haven't close the plot explictly.

File: gradient_descent/gradient_descent.py [hunk 8]
```diff
@@ -102,44 +90,33 @@ def download_data():
...
     biased_tuple = numpy.asmatrix(numpy.ones((1, num_data), dtype=float).transpose())
     train_data = numpy.column_stack((biased_tuple, train_data))
-    output_cols = (11,)
-    train_output = numpy.asmatrix(numpy.loadtxt(home + '/Documents/rock_aquifer_train.dat',
-                                                usecols=output_cols)).transpose()
+    train_output = train_output.transpose()
     parameter_vector = numpy.asmatrix([2, 4, 1, 5, 4, 1, 2, 2, 3, 1, 1, 2]).transpose()
     learning_rate = 0.00015
-    absolute_error_limit = 0.000015
+    absolute_error_limit = 0.000015/4
```
Comment (yashLadha) [inline gradient_descent/gradient_descent.py]: Why this is so? Why you are not using the simplified version?

File: gradient_descent/gradient_descent.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    plt.ylim([-5, 200])
+    plt.legend()
+    plt.show()
+
+
+def download_data():
+    """
+    Downloads test and train data from GitHub repository
+    """
+    import requests
```
Comment (yashLadha) [inline gradient_descent/gradient_descent.py]: Put the import , where all the declarations are defined

**Comment by yashLadha** [comment]
There are conflicts in the `README.md` please resolve them.

**Comment by Prakash2403** [comment]
@yashLadha Please review again.

**Comment by yashLadha** [comment]
@Prakash2403 have u passed a parameter for plotting the graph. ??

File: gradient_descent/gradient_descent.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    :param input_data: Input data to be tested
+    :param output_data: Actual Output data for Input dataset
+    :param parameter_vector: Weight vector after optimized by using gradient descent
+    :return: None
+    """
+    actual_output = output_data
+    hypothesis_output = _hypothesis_value(input_data,
+                                          parameter_vector=parameter_vector)
+    num_examples = len(output_data)
+    plt.plot(range(num_examples), actual_output, 'r', label='Actual Output')
```
Comment (yashLadha) [inline gradient_descent/gradient_descent.py]: Check for argument passed for plotting the graph

**Comment by yashLadha** [comment]
Add matplotlib to requirements. @Prakash2403

**Comment by Prakash2403** [comment]
@yashLadha Please review again.

File: gradient_descent/gradient_descent.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+    """
+    :param input_data: Input data to be tested
+    :param output_data: Actual Output data for Input dataset
+    :param parameter_vector: Weight vector after optimized by using gradient descent
+    :return: None
+    """
+    actual_output = output_data
+    hypothesis_output = _hypothesis_value(input_data,
+                                          parameter_vector=parameter_vector)
+    if len(sys.argv) == 2:
```
Comment (yashLadha) [inline gradient_descent/gradient_descent.py]: Nice Work

**Comment by aviaryan** [comment]
@Prakash2403 Can you please rebase this PR? matplotlib has been already added.

**Comment by Prakash2403** [comment]
@aviaryan Sir, there is an on-going discussion regarding this issue on gitter. So, it is best to not merge the PR right now.

**Comment by Monal5031** [comment]
@Prakash2403 sir please update this PR.

**Comment by singhpratyush** [comment]
Please reopen if you wish to continue.
